[16.0][FIX] fastapi: Avoid zombie threads

lmignon opened the pull request:
Each time a fastapi app is created, a new event loop thread is created by the ASGIMiddleware. Unfortunately, every time the cache is cleared, a new app is created with a new even loop thread. This leads to an increase in the number of threads created to manage the asyncio event loop, even though many of them are no longer in use. To avoid this problem, the thread in charge of the event loop is now created only once per thread / process and the result is stored in the thread's local storage. If a new instance of an app needs to be created following a cache reset, this ensures that the same event loop is reused.

refs #484

lmignon committed:
[FIX] fastapi: Avoid zombie threads

Each time a fastapi app is created, a new event loop thread is created by the ASGIMiddleware. Unfortunately, every time the cache is cleared, a new app is created with a new even loop thread. This leads to an increase in the number of threads created to manage the asyncio event loop, even though many of them are no longer in use. To avoid this problem, the thread in charge of the event loop is now created only once per thread / process and the result is stored in the thread's local storage. If a new instance of an app needs to be created following a cache reset, this ensures that the same event loop is reused.

refs #484

lmignon commented:
ping @sebalix

sebalix commented on fastapi/models/fastapi_endpoint.py:
```diff
@@ -19,6 +21,26 @@
+# Using a thread-local storage allows to have a dedicated event loop per thread
+# and avoid the need to create a new event loop for each request. It's also
+# compatible with the multi-worker mode of Odoo.
+_event_loop_storage = threading.local()
```
<s>Indeed, having one loop per thread is mandatory, missed this in #485 :+1: </s>
EDIT: changing my mind, not working as expected, see below :arrow_down:

sebalix replied:
I still have doubt here: this will attach the loop to the current thread, which is OK in workers mode as the current thread is always the main thread of the worker, but in multi-threads, that would mean we will have one loop per spawned thread to handle the user's request, which is not the main thread. Am I wrong?

I try to test all of this locally.

sbidoul replied:
Oh, is Odoo spawning a new thread for each request in multithread mode?

sebalix replied:
What I did to reproduce this issue:

1. Here is Odoo running in multi-threads (2 exactly, forced with `ODOO_MAX_HTTP_THREADS` + cron threads disabled, and without queue_job neither), so we have the main thread+2 = 3:
![image](https://github.com/user-attachments/assets/c13c30f6-cdae-4e8e-8905-49ed160821cd)

2. Request done on a FastAPI endpoint (e.g. `curl http://localhost:8069/{root_path}/sales`), one daemon thread spawned to handle the loop as expected:
![image](https://github.com/user-attachments/assets/c22c36e0-e19d-49e0-a82a-ca22a420067d)

3. Beside in a Odoo shell, I reset the cache and relaunch a request on FastAPI endpoint:
    ```python
    >>> env.registry.clear_caches()
    >>> env.registry.signal_changes()
    ```
    `curl http://localhost:8069/{root_path}/sales`

4. Repeating step 3. several times: I get a new thread each time, so we are not re-using the loop on requests, while it was working as expected with #485 because having the loop created at the module level was done only once in the main thread, not for each subsequent thread created by Odoo to handle user's requests.

sbidoul replied:
@sebalix does that mean sharing the same event loop across threads? Is that safe?

sebalix replied:
> Oh, is Odoo spawning a new thread for each request in multithread mode?

Yes: https://github.com/odoo/odoo/blob/18.0/odoo/service/server.py#L205-L217

> @sebalix does that mean sharing the same event loop across threads? Is that safe?

I don't know... :man_shrugging: maybe not something supported by default (reading https://www.neilbotelho.com/blog/multithreaded-async.html ).

And what if we attach the event loop to the current thread? Or one not with daemon=True? So we get one event loop created for each user's request, does it imply a lot of overhead? (and we maybe lose the benefit of FastAPI by doing so)

lmignon commented:
@sbidoul additional expert advice would be welcome :smirk:

sbidoul commented on fastapi/models/fastapi_endpoint.py:
```diff
@@ -19,6 +21,26 @@
+        loop_thread = threading.Thread(target=loop.run_forever, daemon=True)
+        loop_thread.start()
+        _event_loop_storage.loop = loop
+        _event_loop_storage.thread = loop_thread
```
Is it necessary to store the loop thread here?

sebalix replied:
IMO no, we don't need it

sbidoul reviewed (dismissed):
Just a minor question, but otherwise this makes sense and looks good to me.

sebalix reviewed (commented):
Not approving based on my test here: https://github.com/OCA/rest-framework/pull/486/files#r1908260365

lmignon commented:
@sbidoul @sebalix 
Following the last comments, I propose to put in place a pool of event loop thread. The pool capacity will be set to the number of workers or to the max number of thread. In the same time, we can override the method  __call__ of the middelware to request the loop at the start of the request processing and release it at the end. In this way we will ensure a proper isolation of the event loop by request and the reuse of the pool in a safe way in multi process and multi thread mode....

Code will folllow

lmignon committed:
[IMP] fastapi: add event loop lifecycle management

This commit adds event loop lifecycle management to the FastAPI dispatcher.

Before this commit, an event loop and the thread to run it were created
each time a FastAPI app was created. The drawback of this approach is that
when the app was destroyed (for example, when the cache of app was cleared),
the event loop and the thread were not properly stopped, which could lead
to memory leaks and zombie threads. This commit fixes this issue by creating
a pool of event loops and threads that are shared among all FastAPI apps.
On each call to a FastAPI app, a event loop is requested from the pool and
is returned to the pool when the app is destroyed. At request time of
an event loop, the pool try to reuse an existing event loop and if no event
loop is available, a new event loop is created.

The cache of the FastAPI app is also refactored to use it's own mechanism.
It's now based on a dictionary of queues by root path by database,
where each queue is a pool of FastAPI app. This allows a better management
of the invalidation of the cache. It's now possible to invalidate
the cache of FastAPI app by root path without affecting the cache of others
root paths.

lmignon commented:
@sebalix @sbidoul I've implemented a new approach based on queue.Queue to manage a dynamic pool of event loops and the fastapi application cache. I'm probably still missing a bit of code for shutting down the threads used for the event pool when the server stops. I haven't yet figured out how to implement this, even though the method is available on the pool to stop all the threads.

sbidoul commented on fastapi/pools/event_loop.py:
```diff
@@ -0,0 +1,68 @@
+            loop = asyncio.new_event_loop()
+            thread = threading.Thread(target=loop.run_forever, daemon=True)
+            thread.start()
+            return loop, thread
```
Should this not be put in the pool?

sbidoul replied:
no

sbidoul commented on fastapi/pools/event_loop.py:
```diff
@@ -0,0 +1,68 @@
+        Get an event loop from the pool. If no event loop is available, create a new one.
+        """
+        try:
+            return self.pool.get_nowait()
```
If max_size is set, should this not be a blocking `get()` instead ? Otherwise if the pool is empty, it will create a new thread and possibly block when putting threads back in the pool when it is full again?

sbidoul replied:
Which is a theoretical concern since `EventLoopPool` does not seem to ever be initialized with a max_size. So perhaps remove support for max_size for the sake of simplicity, knowing that the native Odoo mechanism will limit the number of concurrent threads?

lmignon replied:
I'll remove max size. It's useless and could conflict with the native Odoo mechanism.

lmignon replied:
Moreever, the code below is a reliquat from a first implementation where my idea was to apply any limits set by Odoo to the pool. It was a very bad idea and far too complex.

sbidoul commented on fastapi/pools/fastapi_app.py:
```diff
@@ -0,0 +1,52 @@
+    def __get_app(self, env: Environment, root_path: str) -> FastAPI:
+        db_name = env.cr.dbname
+        try:
+            return self._queue_by_db_by_root_path[db_name][root_path].get_nowait()
```
Is this double defaultdict initialization thread safe?

lmignon replied:
Indeed, the literature says that `defaultdict` is not thread-safe when operations like key creation and modification occur concurrently from multiple threads.... This is now fixed in https://github.com/OCA/rest-framework/pull/486/commits/289868dd3b8b67104560cf2dfdc22671e38b2211

lmignon replied:
@sbidoul Thinking again about the fact that `defautldict` is not thread safe when assigning a value and creating the default value if a key is missing, I'm not sure it's a real problem in this case The aim of the cache here is to provide a `FasAPI` application for processing a request while ensuring that only one application is used for a request at a time and avoiding the creation of a `FasAPI` app for each request (creating an app is a costly operation as it involves analysing all the routes defined in the app, etc.). The only risk/consequence related to the fact that the assignment to the 'defaultdict' is not thread safe is the possible loss of app created at system startup.... In my opinion, this is without consequence.  Even more, I think it's almost impossible for this to happen. The first operation when an app is requested from the cache is to create a queue and assign it to the cache if it doesn't already exist. It's only at the end of the use of this app that it's going to be put in the queue (at this time there is no more assignment to the dict, it's therefore thread safe). In the end I've come to the conclusion that in the worst case scenario we could have created a queue object for nothing but nothing more.  Therefore, the last commit seems to me overkill since introducing an external lock around a queue.Queue can negate many of the advantages of its built-in thread safety if every access to the queue is wrapped in the lock. This approach can lead to unnecessary contention and reduce the performance benefits of queue.Queue's fine-grained, internal synchronization. 
What do you think about?

sbidoul replied:
> worst case scenario we could have created a queue object for nothing but nothing more

I agree. On the other hand, especially with free threaded python around the corner, I'm not that comfortable with multiple thread manipulating the same dictionary.

> introducing an external lock around a queue.Queue

It's not a lock on the queue, it's a lock on the creation of the queue, right?
That should be really cheap and negligible compared to everything else going on.

lmignon committed:
[FIX] fastapi: Graceful shutdown of event loop

On server shutdown, ensure that created the event loops are closed properly.

lmignon committed:
[FIX] fastapi: Ensure thread safety of the FastAPI app cache

defaultdict in python is not thread safe. Since this data structure
is used to store the cache of FastAPI apps, we must ensure that the
access to this cache is thread safe. This is done by using a lock
to protect the access to the cache.

lmignon committed:
[IMP] fastapi: Improves app cache lifecycle

This commit improves the lifecycle of the fastapi app cache.
It first ensures that the cache is effectively invalidated when changes
are made to the app configuration even if theses changes occur into an
other server instance.
It also remove the use of a locking mechanism put in place to ensure a thread
safe access to a value into the cache to avoid potential concurrency issue when
a default value is set to the cache at access time. This lock could lead
to unnecessary contention and reduce the performance benefits of queue.Queue's
fine-grained internal synchronization for a questionable gain. The only
expected gain was to avoid the useless creation of a queue.Queue instance
that would never be used since at the time of puting the value into the cache
we are sure that a value is already present into the dictionary.

lmignon commented:
This PR works in production for the last 6 weeks. I plan to merge-it tomorrow.  ping @AnizR @qgroulard @sebalix

AnizR commented:
> This PR works in production for the last 6 weeks. I plan to merge-it tomorrow. ping @AnizR @qgroulard @sebalix

Thanks for this pr and the ping :smile: 
I'll integrate this into my projects.

Small interrogation: can these 'zombie' threads be an explanation of a slow but steady increase of memory usage?

lmignon commented:
> Small interrogation: can these 'zombie' threads be an explanation of a slow but steady increase of memory usage?

It could be, but it's not the element that led to the detection of this problem, nor even a factor considered in the diagnostic phase.

AnizR commented on fastapi/pools/fastapi_app.py:
```diff
@@ -0,0 +1,130 @@
+        try:
+            return pool.get_nowait()
+        except queue.Empty:
+            env["fastapi.endpoint"].sudo()
```
I think that this line is not necessary and can be removed, right?

lmignon replied:
@AnizR I removed the useless line of code.

AnizR reviewed (changes requested):
Here is my review, thanks for you PR.
I have just one remark.
Besides that, the code seems good to me :+1:

lmignon committed:
[FIX] fastapi: remove useless code

lmignon commented:
/ocabot merge minor
